### DISCUSSION TITLE
High memory usage? Over 200MB for mostly-idle container

### DISCUSSION THREAD

**Discussion by binarydad** [shlinkio/discussions#59] [Help wanted] [answered]
#### How Shlink is set-up

Containerized, version 3.1.2, hosted on ubuntu docker host. Database is mysql on another server. 

#### Summary
Curious as to why there is high-RAM usage. I have 2 instances and both use a little above/below 200MB on average. Wondering why this is. Screenshot below is just shy of 200MB, but I've seen it go past that mark. Thanks!

![image](https://user-images.githubusercontent.com/1760607/184177913-c85004fb-5627-4d11-bd4c-4cd152d72fa3.png)

**Comment by acelaya** [comment]
It's hard to tell, but I imagine [openswoole](https://openswoole.com/) (the runtime used to serve Shlink on the docker image) reserves some memory by default.

I will try to investigate a bit further how this works, and if it's configurable, but if your instances don't have a lot of traffick, maybe you can try reducing the amount of task and web workers, as described [here](https://shlink.io/documentation/environment-variables/#server)

It uses 16 of each by default. I would try with 8 or so and see if it helps.

**Comment by binarydad** [comment]
Awesome. I'll check this out and report back. Thanks!


------ Original Message ------
From "Alejandro Celaya" ***@***.***>
To "shlinkio/shlink" ***@***.***>
Cc "Ryan Peters" ***@***.***>; "Author" 
***@***.***>
Date 8/11/2022 2:55:31 PM
Subject Re: [shlinkio/shlink] High memory usage? Over 200MB for 
mostly-idle container (Discussion shlinkio/shlink#1505)

>
>It's hard to tell, but I imagine openswoole <https://openswoole.com/> 
>(the runtime used to serve Shlink on the docker image) reserves some 
>memory by default.
>
>I will try to investigate a bit further how this works, and if it's 
>configurable, but if your instances don't have a lot of traffick, maybe 
>you can try reducing the amount of task and web workers, as described 
>here <https://shlink.io/documentation/environment-variables/#server>
>
>It uses 16 of each by default. I would try with 8 or so and see if it 
>helps.
>
>—
>Reply to this email directly, view it on GitHub 
><https://github.com/shlinkio/shlink/discussions/1505#discussioncomment-3378284>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AANN2X7DLXX2H3UIPEPT5BDVYVECHANCNFSM56JGWNHQ>.
>You are receiving this because you authored the thread.Message ID: 
***@***.***>
>

**Comment by webysther** [comment]
Will be nice to see something in this area, the memory usage for the api in really high when is IDLE for almost 99% of time:

<img width="1212" height="729" alt="image" src="https://github.com/user-attachments/assets/b2eec093-5078-4484-9242-85fcf60fbee9" />

**Reply by acelaya** [reply]
This is a very old thread. Shlink is no longer using openswoole, but RoadRunner.

RoadRunner will reserve around 50/60 MB of memory per worker, so the best approach to reduce that is reduce the amount of workers if you are not having a lot of traffic.

In any case, 250MB doesn't seem terrible to me 😅

**Reply by acelaya** [reply]
I was investigating what looked like a memory leak some time ago. You can see my notes there https://github.com/shlinkio/shlink/issues/2433#issuecomment-3232170268

**Reply by webysther** [reply]
Yeap, I see here https://github.com/roadrunner-server/roadrunner/blob/3c6ee7d781f5d3f3f9d0678fc4c1a1719d6167f8/.rr.yaml#L340 will use a bit of memory for worker. TBH 160MB is high for IDLE, look how amazing the admin are using only 5MB.

**Reply by webysther** [reply, marked as answer]
After change this:

```yml
WEB_WORKER_NUM=1
TASK_WORKER_NUM=1
```

Worked, I think this can be close!

**Reply by acelaya** [reply]
> Yeap, I see here https://github.com/roadrunner-server/roadrunner/blob/3c6ee7d781f5d3f3f9d0678fc4c1a1719d6167f8/.rr.yaml#L340 will use a bit of memory for worker.

That's only the maximum memory you allow for a worker before it gets killed, not the minimum. 

Shlink would require an in-depth memory profiling to find out where is the memory going. Would be great if you want to give that a chance.

> TBH 160MB is high for IDLE, look how amazing the admin are using only 5MB.

It makes no sense to compare those two really.

**Reply by webysther** [reply]
Good point about the memory, I will try use this to force some SWAP as part of this is a possible dirt pages:

```yml
deploy:
      resources:
        limits:
          cpus: '2.00'
          memory: 256M
```

Last time I worked with PHP was 5.3-7-ish
